File: src/lib/qcr/fair.js
// FAIR decomposition and deterministic expected loss — port of v0 fair.py.
import { pertMean } from '@/lib/qcr/estimates';

// Factor means and annualized loss expectancy.
// LEF = TEF × Vulnerability; magnitude = primary + SL·SLP; ALE = LEF × magnitude.
export function expectedLoss(fair) {
  const tef = pertMean(fair.threat_event_frequency);
  const vulnerability = pertMean(fair.vulnerability);
  const lef = tef * vulnerability;
  const primaryLoss = pertMean(fair.primary_loss);
  const expectedSecondaryLoss = pertMean(fair.secondary_loss) * pertMean(fair.secondary_loss_probability);
  const lossMagnitude = primaryLoss + expectedSecondaryLoss;
  return {
    tef,
    vulnerability,
    lef,
    primaryLoss,
    expectedSecondaryLoss,
    lossMagnitude,
    ale: lef * lossMagnitude,
  };
}

// Display-ready FAIR inputs. labelKey resolves through i18n at render time.
export function decompositionRows(fair) {
  return [
    { labelKey: 'fair.factorTef', estimate: fair.threat_event_frequency },
    { labelKey: 'fair.factorVulnerability', estimate: fair.vulnerability },
    { labelKey: 'fair.factorPrimaryLoss', estimate: fair.primary_loss },
    { labelKey: 'fair.factorSecondaryLoss', estimate: fair.secondary_loss },
    { labelKey: 'fair.factorSlp', estimate: fair.secondary_loss_probability },
  ];
}

File: qcr_core/fair.py
"""FAIR decomposition and deterministic expected-loss calculations."""

from __future__ import annotations

from dataclasses import dataclass

from .estimates import pert_mean
from .models import FairModel


@dataclass(frozen=True)
class ExpectedLossResult:
    threat_event_frequency: float
    vulnerability: float
    loss_event_frequency: float
    primary_loss: float
    expected_secondary_loss: float
    loss_magnitude: float
    annualized_loss_expectancy: float


def expected_loss(model: FairModel) -> ExpectedLossResult:
    """Calculate FAIR factor means and annualized loss expectancy."""
    tef = pert_mean(model.threat_event_frequency)
    vulnerability = pert_mean(model.vulnerability)
    lef = tef * vulnerability
    primary = pert_mean(model.primary_loss)
    secondary = pert_mean(model.secondary_loss) * pert_mean(model.secondary_loss_probability)
    magnitude = primary + secondary
    return ExpectedLossResult(tef, vulnerability, lef, primary, secondary, magnitude, lef * magnitude)


def decomposition_rows(model: FairModel) -> list[dict[str, float | str]]:
    """Return display-ready FAIR inputs without coupling calculations to Streamlit."""
    factors = {
        "Threat Event Frequency": model.threat_event_frequency,
        "Vulnerability": model.vulnerability,
        "Primary Loss": model.primary_loss,
        "Secondary Loss": model.secondary_loss,
        "Secondary Loss Probability": model.secondary_loss_probability,
    }
    return [
        {
            "Factor": name,
            "Minimum": value.minimum,
            "Most likely": value.most_likely,
            "Maximum": value.maximum,
            "Unit": value.unit,
        }
        for name, value in factors.items()
    ]

